### PR TITLE
Close obs context manually after adding status code

### PR DESCRIPTION
**Pull request by tharinduwijewardane**
## Purpose
Manually closing the observer context after adding status code to resource span. It is required when the user has put a return statement instead of a caller->respond method in the resource function.

Lang PR: https://github.com/ballerina-platform/ballerina-lang/pull/30074

## Checklist
- [x] Linked to an issue
- [ ] Updated the changelog
- [ ] Added tests -- will be added with https://github.com/wso2-enterprise/choreo/issues/3437

### CODE REVIEW

File: http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java [hunk 1]
```diff
@@ -47,6 +51,11 @@
     public void notifySuccess(Object result) {
         if (result == null) { // handles nil return and end of resource exec
             requestMessage.waitAndReleaseAllEntities();
+            ObserverContext observerContext
```
Comment (chamil321) [inline http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java:54]: Shouldn't we check whether `ObserveUtils.isObservabilityEnabled()` before looking for the `ObserverContext`?

Comment (tharinduwijewardane) [reply http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java:54]: `ObserverContext` will be null when observability is disabled hence that check is redundant. Should I add it anyway?

Comment (chamil321) [reply http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java:54]: Yeah, we can get rid of the map lookup and the cast if check it right? Shall we do that by considering the potential perf gain?

Comment (tharinduwijewardane) [reply http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java]: Fixed

**Review by chamil321 — APPROVED** [review]

File: http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java [hunk 2]
```diff
@@ -47,6 +51,13 @@
     public void notifySuccess(Object result) {
         if (result == null) { // handles nil return and end of resource exec
             requestMessage.waitAndReleaseAllEntities();
+            if (ObserveUtils.isObservabilityEnabled()) {
+                ObserverContext observerContext
+                        = (ObserverContext) requestMessage.getProperty(OBSERVABILITY_CONTEXT_PROPERTY);
+                if (observerContext != null) {
+                    ObserveUtils.stopObservationWithContext(observerContext);
```
Comment (nadundesilva) [inline http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java:58]: Don't we need to check whether the observation should be manually closed (`if (observerContext.isManuallyClosed())`) here before closing it ? (Same in the `sendFailureResponse` method as well)

Comment (tharinduwijewardane) [reply http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java:58]: `observerContext` being present in the request message means it should be manually closed. Should I explicitly check it?

Comment (nadundesilva) [reply http-native/src/main/java/org/ballerinalang/net/http/HttpCallableUnitCallback.java:58]: Ahh yes. It should be enough :+1:

**Review by nadundesilva — APPROVED** [review]

**Pull request merged**